typescript/pragma-utils/src/starknet/gas.ts

```diff
@@ -0,0 +1,295 @@
+    keccak: rawResources.keccak_builtin_applications ?? 0,
+    bitwise: rawResources.bitwise_builtin_applications ?? 0,
+    ec_op: rawResources.ec_op_builtin_applications ?? 0,
+  };
+
+  const blockNumber = receipt.block_number;
+  const blockInfo = await provider.getBlockWithReceipts(blockNumber);
+
+  const stateUpdate = await provider.getStateUpdate(blockNumber);
+  const storageDiffs = stateUpdate.state_diff.storage_diffs;
+  const paidInStrk = receipt.actual_fee.unit == "FRI";
+  const gasPrice = BigInt(
+    paidInStrk
+      ? blockInfo.l1_gas_price.price_in_fri
+      : blockInfo.l1_gas_price.price_in_wei,
+  );
+
+  const gasPerComputationCategory = Object.fromEntries(
+    Object.entries(executionResources)
+      .filter(([resource]) => resource in gasWeights)
+      .map(([resource, usage]) => [
+        resource,
+        Math.ceil(usage * gasWeights[resource]),
+      ]),
+  );
+  const maxComputationCategory = maxBy(
+    Object.entries(gasPerComputationCategory),
+    ([, gas]) => gas,
+  )![0];
+  const computationGas = BigInt(
+    gasPerComputationCategory[maxComputationCategory],
+  );
+
+  let gasWithoutDa;
+  let feeWithoutDa;
+  let daFee;
+  if (rawResources.data_availability) {
+    const dataGasPrice = Number(
+      paidInStrk
+        ? blockInfo.l1_data_gas_price.price_in_fri
+        : blockInfo.l1_data_gas_price.price_in_wei,
+    );
+
+    daFee =
+      (rawResources.data_availability.l1_gas +
+        rawResources.data_availability.l1_data_gas) *
+      dataGasPrice;
+    feeWithoutDa = actualFee - BigInt(daFee);
+    gasWithoutDa = feeWithoutDa / gasPrice;
+  } else {
+    // This only happens for tx before Dencun
+    gasWithoutDa = actualFee / gasPrice;
+    daFee = gasWithoutDa - computationGas;
+    feeWithoutDa = actualFee;
+  }
+
+  const sortedResources = Object.fromEntries(
+    sortBy(Object.entries(executionResources), 0),
+  );
+
+  // L2 payloads
+  const { calldata, signature } = (await provider.getTransaction(
+    receipt.transaction_hash,
+  )) as any;
+  const calldataGas =
+    calldata && signature
+      ? Math.floor(
+          (calldata.length + signature.length) * l2PayloadsWeights.calldata,
+        )
+      : undefined; // TODO find workaround for deployment transactions
+
+  const eventGas = Math.floor(
+    receipt.events.reduce(
+      (sum: any, { keys, data }: { keys: any; data: any }) =>
+        sum +
+        keys.length * l2PayloadsWeights.eventKey +
+        data.length * l2PayloadsWeights.eventData,
+      0,
+    ),
+  );
+  return {
+    actualFee,
+    paidInStrk,
+    gasWithoutDa,
+    feeWithoutDa,
+    daFee,
+    computationGas,
+    maxComputationCategory,
+    gasPerComputationCategory,
+    executionResources: sortedResources,
+    gasPrice,
+    storageDiffs,
+    daMode: blockInfo.l1_da_mode,
+    calldataGas,
+    eventGas,
+  };
+}
+
+type Profile = Awaited<ReturnType<typeof profileGasUsage>>;
+
+export function newProfiler(provider: RpcProvider) {
```

Comment:
where is that used ?

Reply:
Nowhere yet - just grabbed it from github, might be useful *one day*